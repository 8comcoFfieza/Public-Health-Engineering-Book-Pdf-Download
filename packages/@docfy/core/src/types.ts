import { Node } from 'unist';
import { Processor, Plugin, Settings } from 'unified';

export interface Heading {
  title: string;
  id: string;
  depth: number;
  headings?: Heading[];
}

export interface Page {
  source: string;
  url: string;
  editUrl: string;
  title: string;
  headings: Heading[];
  metadata: Record<string, unknown>;
}

export interface PageContent extends Page {
  ast: Node;
  markdown: string;
  rendered: string;
  demos?: PageContent[];
}

interface ContextOptions
  extends Omit<Options, 'remarkPlugins' | 'tocMaxDepth'> {
  tocMaxDepth: number;
}

export interface Context {
  remark: Processor;
  pages: PageContent[];
  options: ContextOptions;
}

export interface NestedRuntimeOutput {
  name: string;
  label: string;
  pages: Page[];
  children: NestedRuntimeOutput[];
}

export interface RuntimeOutput {
  flat: Page[];
  nested: NestedRuntimeOutput;
}

export interface SourceSettings {
  /**
   * The absolute path to where the files are located.
   */
  root: string;

  /**
   * Match files using the patterns the shell uses, like stars and stuff. It
   * uses Glob package.
   */
  pattern: string;

  /**
   * Pattern to ignore.
   */
  ignore?: string[];

  /**
   * Option to change how urls are generated.
   *
   * There are two ways URLs can ge generated.
   *
   * 1. "auto": Uses the folder structure to inform how the urls are generated.
   *    For example, if you have the folowing files:
   *      - install.md
   *      - components/
   *        - button.md
   *        - card.md
   *    The urls would look like this, (assuming urlPrefix is set to "docs").
   *      - docs/install
   *      - docs/components/buttons
   *      - docs/components/card
   *
   * 2. "manual": It uses frontmatter information to inform "subcategory" and
   *    "category" of the file, ignoring the original file location.
   *    The url schema is as follows:
   *    {category}/{subcategory}/{file-name}
   *
   *    If no category or subcategory is specified, all files will be at the root
   *    level.
   *
   *    This options is perfect for documenting monorepo projects while keeping
   *    documentation next to implementation.
   */
  urlSchema?: 'auto' | 'manual';

  /**
   * Option to prefix the urls.
   *
   * For example:
   * "docs" will generate urls like "/docs/something"
   * "blog" will generate urls like "/blog/something"
   */
  urlPrefix?: string;

  /**
   * Option to suffix the urls.
   *
   * For example:
   * ".html" will generate urls like "/something.html"
   */
  urlSuffix?: string;

  /**
   * Overwrite repository config for this source.
   */
  repository?: RepositoryConfig;
}

interface RepositoryConfig {
  /**
   *
   * The url to the Git Repository
   * Example: https://github.com/josemarluedke/docfy
   */
  url: string;

  /**
   * Branch used to edit your markdown when clicking on "Edit this page" button.
   * @default "master"
   */
  editBranch?: string;
}

export interface Options {
  /**
   * Additional remark plugins
   *
   * Example:
   *
   * ```js
   * const hbs = require('remark-hbs');
   * const autolinkHeadings = require('remark-autolink-headings');
   *
   * const remarkPlugins = [
   *   [
   *     autolinkHeadings,
   *     {
   *       behavior: 'wrap'
   *     }
   *   ],
   *   hbs
   * ];
   * ```
   */
  remarkPlugins?: ([Plugin, Settings] | Plugin)[];

  /**
   * The max depth of headings
   * @default 6
   */
  tocMaxDepth?: number;

  /**
   * The repository config
   */
  repository?: RepositoryConfig;
}

interface DocfyConfigSourceSettings extends Omit<SourceSettings, 'root'> {
  root?: string;
}

export interface DocfyConfig extends Options {
  sources: DocfyConfigSourceSettings[];
  labels?: Record<string, string>;
}
