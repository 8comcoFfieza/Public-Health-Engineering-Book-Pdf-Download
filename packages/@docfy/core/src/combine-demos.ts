import path from 'path';
import { Page, Context } from './types';

/*
 * Finds the index of the owner for a given demo file.
 *
 * Demos should be located in a folder named `demo/` or `something-demo/`.
 *
 * eg.
 * Below the owner would be the button/index.md:
 * components/button/
 *   - index.md
 *   - demo/
 *     - demo1.md
 *     - demo2.md
 *
 * Below the owner would be the button.md:
 * components/
 *   - button.md
 *   - button-demo/
 *     - demo1.md
 *     - demo2.md
 */
function findDemoOwner(contents: Page[], demoSource: string): number {
  const folder = path.basename(path.dirname(demoSource));

  let parentName = folder.replace('-demo', '');
  if (parentName === 'demo') {
    parentName = path.basename(path.dirname(path.dirname(demoSource)));
  }

  return contents.findIndex((item: Page): boolean => {
    const file = path.parse(path.basename(item.source));
    return (
      file.name === parentName ||
      (file.name === 'index' &&
        path.basename(path.dirname(item.source)) === parentName)
    );
  });
}

export default function CombineDemos(context: Context): Context {
  context.pages.forEach((item: Page, index: number): void => {
    const folder = path.basename(path.dirname(item.source));

    if (folder.match(/demo$/)) {
      const parent =
        context.pages[findDemoOwner(context.pages, item.source)];

      if (parent) {
        if (!parent.demos) {
          parent.demos = [item];
        } else {
          parent.demos.push(item);
        }

        context.pages.splice(index, 1);
      }
    }
  });

  return context;
}
