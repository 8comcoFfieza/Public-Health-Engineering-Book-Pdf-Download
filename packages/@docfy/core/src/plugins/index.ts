export { fixUrls } from './fix-urls';
export { combineDemos } from './combine-demos';
