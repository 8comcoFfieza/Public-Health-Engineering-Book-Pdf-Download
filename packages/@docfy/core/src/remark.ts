import unified from 'unified';
import parse from 'remark-parse';
import frontmatter from 'remark-frontmatter';
import slug from 'remark-slug';
import html from 'remark-html';
import stringify from 'remark-stringify';
import normalizeHeadings from 'remark-normalize-headings';
import { Processor } from 'unified';
import { Options } from './types';

export function createRemark(plugins?: Options['remarkPlugins']): Processor {
  const stack = unified()
    .use(parse)
    .use(normalizeHeadings)
    .use(slug)
    .use(stringify)
    .use(frontmatter, [{ type: 'yaml', marker: '-' }]);

  if (plugins && plugins.length > 0) {
    plugins.forEach((fn) => {
      if (Array.isArray(fn)) {
        stack.use(...fn);
      } else {
        stack.use(fn);
      }
    });
  }

  return stack.use(html);
}
