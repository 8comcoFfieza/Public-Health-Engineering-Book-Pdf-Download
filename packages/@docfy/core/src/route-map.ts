import { Content } from './types';

export default function routeMap(content: Content[], prefix = ''): string[] {
  const routes: string[] = [];

  // intro, bla, category: [bla, bla], package: [bla, bla, category]

  content.forEach((item) => {
    if (item.metadata.url) {
      routes.push(prefix + item.metadata.url);
    }
  });

  return routes;
}
