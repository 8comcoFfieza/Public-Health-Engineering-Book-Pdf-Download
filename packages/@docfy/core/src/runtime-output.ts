import { Page, PageContent, RuntimeOutput, NestedRuntimeOutput } from './types';

function findChild(
  node: NestedRuntimeOutput,
  name: string
): NestedRuntimeOutput | undefined {
  return node.children.find((item) => {
    return item.name === name;
  });
}

function sortByOrder(pages: Page[]): Page[] {
  return pages.sort((a, b) => {
    const aOrder =
      typeof a.metadata.order !== 'undefined' ? Number(a.metadata.order) : 998;
    const bOrder =
      typeof b.metadata.order !== 'undefined' ? Number(b.metadata.order) : 999;
    return aOrder - bOrder;
  });
}

export function genereateFlatOutput(pageContents: PageContent[]): Page[] {
  const pages: Page[] = pageContents.map((page) => {
    const { url, headings, title, source, metadata } = page;

    return {
      url,
      headings,
      title,
      source,
      metadata
    };
  });

  return sortByOrder(pages);
}

export function generateNestedOutput(
  pages: Page[],
  labels: Record<string, string> = {},
  existingObj?: NestedRuntimeOutput
): NestedRuntimeOutput {
  const node: NestedRuntimeOutput = existingObj || {
    name: '/',
    label: labels['/'] || '/',
    pages: [],
    children: []
  };

  pages.forEach((item): void => {
    let url = (item.metadata.relativeUrl as string) || item.url;
    url = url[0] === '/' ? url.substring(1) : url;
    const urlParts = url.split('/');

    if (urlParts.length === 1) {
      item.metadata.relativeUrl = urlParts[0];
      node.pages.push(item);
    } else {
      const name = urlParts.shift();

      if (typeof name === 'string') {
        let child = findChild(node, name);

        if (!child) {
          child = {
            name: name,
            label: labels[name] || name,
            pages: [],
            children: []
          };
          node.children.push(child);
        }

        item.metadata.relativeUrl = urlParts.join('/');
        generateNestedOutput([item], labels, child);

        sortByOrder(child.pages);
      }
    }
  });

  sortByOrder(node.pages);
  return node;
}

export function generateRuntimeOutput(
  pageContents: PageContent[],
  labels: Record<string, string> = {}
): RuntimeOutput {
  const flat = genereateFlatOutput(pageContents);
  const nested = generateNestedOutput(flat, labels);

  return { flat, nested };
}
