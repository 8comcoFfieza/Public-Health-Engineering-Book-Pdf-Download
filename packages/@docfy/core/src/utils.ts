import path from 'path';
import visit from 'unist-util-visit';
import { Node } from 'unist';
import { Page } from './types';
import toString from 'mdast-util-to-string';
import Slugger from 'github-slugger';
import YAML from 'yaml';
import url from 'url';

const slug = Slugger.slug;

export function generateManualUrl(
  source: string,
  meta: Page['metadata'],
  prefix?: string,
  suffix?: string
): string {
  const parts: string[] = [''];
  if (prefix) {
    parts.push(prefix);
  }
  if (meta.package) {
    parts.push(slug(meta.package));
  }
  if (meta.category) {
    parts.push(slug(meta.category));
  }
  let fileName = path.parse(path.basename(source)).name;
  if (fileName === 'index') {
    fileName = path.basename(path.dirname(source));
  }
  parts.push(`${fileName}${suffix || ''}`);
  return parts.join('/');
}

export function generateAutoUrl(
  source: string,
  prefix?: string,
  suffix?: string
): string {
  const parts: string[] = [''];
  if (prefix) {
    parts.push(prefix);
  }
  parts.push(...source.split(path.sep));

  let fileName = path.parse(parts.pop() as string).name;
  if (fileName === 'index') {
    fileName = path.basename(path.dirname(source));
    parts.pop(); // remove duplicated folder name
  }

  parts.push(`${fileName}${suffix || ''}`);
  return parts.join('/');
}

export function inferTitle(ast: Node): string | undefined {
  let docTitle: string | undefined;
  visit(ast, 'heading', (node) => {
    const { depth } = node;
    if (depth !== 1) return;
    docTitle = toString(node);
  });
  return docTitle;
}

export function parseFrontmatter(source: string, ast: Node): object {
  let result = {};
  visit(ast, 'yaml', (node) => {
    try {
      result = YAML.parse(node.value as string);
    } catch (e) {
      console.error(`Error while parsing frontmatter in ${source}: `, e);
    }
  });

  return result;
}

export function isValidUrl(s: string): boolean {
  try {
    new url.URL(s);
    return true;
  } catch {
    return false;
  }
}

export function isAnchorUrl(s: string): boolean {
  return s[0] === '#';
}
