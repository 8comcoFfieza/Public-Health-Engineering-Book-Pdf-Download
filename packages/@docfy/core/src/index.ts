import glob from 'fast-glob';
import fs from 'fs';
import path from 'path';
import trough, { Through } from 'trough';
import {
  PageContent,
  Context,
  Options,
  SourceConfig,
  PageMetadata,
  DocfyResult
} from './types';
import {
  DEFAULT_IGNORE,
  generateAutoUrl,
  generateManualUrl,
  inferTitle,
  parseFrontmatter
} from './-private/utils';
import { createRemark } from './-private/remark';
import {
  combineDemos,
  renderMarkdown,
  replaceInternalLinks,
  staticAssets,
  toc,
  uniquefyUrls
} from './plugins';
import { getRepoEditUrl } from './-private/repo-info';
import { transformToNestedPageMetadata } from './-private/nested-page-metadata';

export default class Docfy {
  private pipeline: Through<Context>;
  private context: Context;

  constructor(options: Options = {}) {
    const { remarkPlugins, ...rest } = options;
    this.context = {
      remark: createRemark(remarkPlugins),
      pages: [],
      staticAssets: [],
      options: {
        ...rest,
        tocMaxDepth: rest.tocMaxDepth || 6
      }
    };

    this.pipeline = trough<Context>()
      .use<SourceConfig[]>(this.initializePipeline.bind(this))
      .use(combineDemos)
      .use(uniquefyUrls)
      .use(replaceInternalLinks)
      .use(staticAssets);

    if (Array.isArray(options.plugins)) {
      options.plugins.forEach((item) => {
        this.pipeline.use(item);
      });
    }

    // Make sure TOC and renderMarkdown plugins are the last ones
    this.pipeline.use(toc).use(renderMarkdown);
  }

  public run(sources: SourceConfig[]): Promise<DocfyResult> {
    return new Promise((resolve, reject) => {
      this.pipeline.run(sources, (err: unknown, ctx: Context): void => {
        if (err) {
          reject(err);
        } else {
          resolve({
            content: ctx.pages,
            staticAssets: ctx.staticAssets,
            nestedPageMetadata: transformToNestedPageMetadata(
              ctx.pages.map((p) => p.meta),
              ctx.options.labels
            )
          });
        }
      });
    });
  }

  private initializePipeline(sources: SourceConfig[]): Context {
    const ctx = this.context;

    sources.forEach((item) => {
      let repoEditUrl: string | null;

      if (item.repository || ctx.options.repository) {
        repoEditUrl = getRepoEditUrl(
          item.root,
          item.repository?.url || ctx.options.repository?.url || '',
          item.repository?.editBranch || ctx.options.repository?.editBranch
        );
      }
      const files = glob.sync(item.pattern, {
        cwd: item.root,

        ignore: [...DEFAULT_IGNORE, ...(item.ignore || [])],
        absolute: true
      });

      files.forEach((file) => {
        ctx.pages.push(this.createPage(item, file, repoEditUrl));
      });
    });

    return ctx;
  }

  private createPage(
    sourceConfig: SourceConfig,
    fullPath: string,
    repoEditUrl?: string | null
  ): PageContent {
    const relativePath = fullPath.replace(
      path.join(sourceConfig.root, '/'),
      ''
    );

    const markdown = fs.readFileSync(fullPath).toString();
    const ast = this.context.remark.runSync(
      this.context.remark.parse(markdown)
    );
    const frontmatter = parseFrontmatter(fullPath, ast);

    let url: string;
    let title = inferTitle(ast);
    if (typeof frontmatter.title === 'string') {
      title = frontmatter.title;
    }
    // Fallback for title
    if (typeof title === 'undefined' || title === '') {
      title = path.parse(fullPath).name;
    }

    if (typeof frontmatter.url === 'string') {
      url = generateAutoUrl(
        frontmatter.url,
        sourceConfig.urlPrefix,
        sourceConfig.urlSuffix
      );
    } else if (sourceConfig.urlSchema === 'manual') {
      url = generateManualUrl(
        relativePath,
        frontmatter,
        sourceConfig.urlPrefix,
        sourceConfig.urlSuffix
      );
    } else {
      url = generateAutoUrl(
        relativePath,
        sourceConfig.urlPrefix,
        sourceConfig.urlSuffix
      );
    }

    // Add fallback order for index pages
    if (
      url.length > 0 &&
      url[url.length - 1] === '/' &&
      typeof frontmatter.order === 'undefined'
    ) {
      (frontmatter.order as undefined | number) = -1;
    }

    let editUrl = '';

    if (repoEditUrl) {
      editUrl = repoEditUrl.replace('{filepath}', relativePath);
    }

    const metadata: PageMetadata = {
      url,
      title,
      editUrl,
      relativeUrl: undefined,
      headings: [],
      frontmatter: frontmatter,
      pluginData: {}
    };

    return {
      meta: metadata,
      source: relativePath,
      sourceConfig,
      ast,
      markdown,
      rendered: '',
      pluginData: {}
    };
  }
}
