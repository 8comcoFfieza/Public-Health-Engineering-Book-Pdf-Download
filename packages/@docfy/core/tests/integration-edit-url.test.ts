import Docfy from '../src';
import { PageContent } from '../src/types';
import path from 'path';

const root = path.resolve(__dirname, './__fixtures__/monorepo');

describe('Generates Edit Url', () => {
  let pages: PageContent[];

  beforeAll(async () => {
    const docfy = new Docfy({
      repository: {
        url: 'https://github.com/user/repo',
        editBranch: 'dev'
      }
    });
    pages = await docfy.run([
      {
        root,
        urlPrefix: 'docs',
        urlSchema: 'manual',
        pattern: '/**/*.md',
        ignore: ['/**/package2/**/*.md']
      },
      {
        root,
        urlPrefix: 'docs',
        urlSchema: 'manual',
        pattern: '/**/package2/**/*.md',
        repository: {
          url: 'https://bitbucket.org/user/repo'
        }
      }
    ]);
  });

  test('it should generated edit urls correctly', async () => {
    const urls = [];
    pages.forEach((page) => {
      urls.push([page.source, page.editUrl]);
    });

    expect(urls).toMatchSnapshot();
  });
});
