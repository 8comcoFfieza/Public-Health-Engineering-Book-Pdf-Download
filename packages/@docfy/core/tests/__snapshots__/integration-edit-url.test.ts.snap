// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generates Edit Url it should generated edit urls correctly 1`] = `
Array [
  Array [
    "docs/about.md",
    "https://github.com/user/repo/edit/dev/packages/@docfy/core/tests/__fixtures__/monorepo/docs/about.md",
  ],
  Array [
    "docs/how-it-works.md",
    "https://github.com/user/repo/edit/dev/packages/@docfy/core/tests/__fixtures__/monorepo/docs/how-it-works.md",
  ],
  Array [
    "packages/package1/components/button.md",
    "https://github.com/user/repo/edit/dev/packages/@docfy/core/tests/__fixtures__/monorepo/packages/package1/components/button.md",
  ],
  Array [
    "packages/package1/components/form/index.md",
    "https://github.com/user/repo/edit/dev/packages/@docfy/core/tests/__fixtures__/monorepo/packages/package1/components/form/index.md",
  ],
  Array [
    "packages/package1/docs/intro.md",
    "https://github.com/user/repo/edit/dev/packages/@docfy/core/tests/__fixtures__/monorepo/packages/package1/docs/intro.md",
  ],
  Array [
    "packages/package2/docs/overview.md",
    "https://bitbucket.org/user/repo/src/dev/packages/@docfy/core/tests/__fixtures__/monorepo/packages/package2/docs/overview.md?mode=edit&spa=0&at=dev&fileviewer=file-view-default",
  ],
  Array [
    "packages/package2/docs/styles.md",
    "https://bitbucket.org/user/repo/src/dev/packages/@docfy/core/tests/__fixtures__/monorepo/packages/package2/docs/styles.md?mode=edit&spa=0&at=dev&fileviewer=file-view-default",
  ],
]
`;
